module/module.go

```diff
@@ -0,0 +1,422 @@
+}
+
+// AddComponent receives the component configuration from the parent.
+func (m *Module) AddComponent(ctx context.Context, req *pb.AddComponentRequest) (*pb.AddComponentResponse, error) {
+	cfg, err := config.ComponentConfigFromProto(req.Config)
+	if err != nil {
+		return nil, err
+	}
+	deps := make(registry.Dependencies)
+	for _, c := range req.Dependencies {
+		name, err := resource.NewFromString(c)
+		if err != nil {
+			return nil, err
+		}
+		c, err := m.GetParentComponent(ctx, name)
+		if err != nil {
+			return nil, err
+		}
+		deps[name] = c
+	}
+
+	creator := registry.ComponentLookup(cfg.ResourceName().Subtype, cfg.Model)
+	if creator != nil && creator.Constructor != nil {
+		comp, err := creator.Constructor(ctx, deps, *cfg, m.logger)
+		if err != nil {
+			return nil, err
+		}
+
+		subSvc, ok := m.services[cfg.ResourceName().Subtype]
+		if !ok {
+			return nil, errors.Errorf("module can't service api: %s", cfg.ResourceName().Subtype)
+		}
+
+		err = subSvc.Add(cfg.ResourceName(), comp)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &pb.AddComponentResponse{}, nil
+}
+
+// ReconfigureComponent receives the component configuration from the parent.
+func (m *Module) ReconfigureComponent(ctx context.Context, req *pb.ReconfigureComponentRequest) (*pb.ReconfigureComponentResponse, error) {
+	cfg, err := config.ComponentConfigFromProto(req.Config)
+	if err != nil {
+		return &pb.ReconfigureComponentResponse{}, err
+	}
+
+	svc, ok := m.services[cfg.ResourceName().Subtype]
+	if !ok {
+		return &pb.ReconfigureComponentResponse{}, errors.Errorf("no component service for %+v", cfg.ResourceName())
+	}
+	comp := svc.Resource(cfg.ResourceName().Name)
+
+	deps := make(registry.Dependencies)
+	for _, c := range req.Dependencies {
+		name, err := resource.NewFromString(c)
+		if err != nil {
+			return &pb.ReconfigureComponentResponse{}, err
+		}
+		c, err := m.GetParentComponent(ctx, name)
+		if err != nil {
+			return &pb.ReconfigureComponentResponse{}, err
+		}
+		deps[name] = c
+	}
+
+	// Check if component directly supports reconfiguration.
+	rc, ok := comp.(registry.ReconfigurableComponent)
+	if ok {
+		return &pb.ReconfigureComponentResponse{}, rc.Reconfigure(ctx, *cfg, deps)
+	}
+
+	// If it can't reconfigure, replace it.
+	err = utils.TryClose(ctx, comp)
+	if err != nil {
+		m.logger.Error(err)
+	}
+	creator := registry.ComponentLookup(cfg.ResourceName().Subtype, cfg.Model)
+	if creator != nil && creator.Constructor != nil {
+		comp, err := creator.Constructor(ctx, deps, *cfg, m.logger)
+		if err != nil {
+			return &pb.ReconfigureComponentResponse{}, err
+		}
+		return &pb.ReconfigureComponentResponse{}, svc.ReplaceOne(cfg.ResourceName(), comp)
+	}
+
+	return &pb.ReconfigureComponentResponse{}, errors.Errorf("can't recreate resource %+v", cfg)
+}
+
+// AddService receives the service configuration from the parent.
+func (m *Module) AddService(ctx context.Context, req *pb.AddServiceRequest) (*pb.AddServiceResponse, error) {
```

Comment:
AddService is nearly identical to AddComponent. The only reason it's different is that config.Component and config.Service are nominally different types (though contain pretty much the same structure these days) and because the service constructor gets the entire parent robot, instead of dependencies. Once service deps are added in RDK, I think we should consider merging components/services as just "resources" and vastly simplify things, including here.